File: frontend1/src/api/analytics.ts
import { apiClient } from './client';
import { UsageLogResponse } from './usage';

export interface EquipmentUtilization {
  equipment_id: string;
  equipment_type: string;
  site_id: string | null;
  site_name: string | null;
  operator_id: string | null;
  operator_name: string | null;
  rental_status: string;
  engine_hours: number | string;
  idle_hours: number | string;
  total_usage_hours: number | string;
  fuel_used: number | string;
  utilization_rate: number;
  idle_percentage: number;
  insight_summary: string;
}

export interface SiteAnalytics {
  site_id: string;
  site_name: string;
  location: string;
  equipment_count: number;
  active_rentals: number;
  total_engine_hours: number | string;
  total_idle_hours: number | string;
  total_fuel_used: number | string;
  average_utilization: number;
}

export interface DailyUsagePoint {
  date: string;
  engine_hours: number;
  idle_hours: number;
  fuel_used: number;
  utilization_rate: number;
}

export interface EquipmentPerformance {
  equipment_id: string;
  equipment_type: string;
  status: string;
  site_id: string | null;
  site_name: string | null;
  operator_id: string | null;
  operator_name: string | null;
  rental_id: string | null;
  checkout_date: string | null;
  expected_return: string | null;
  total_engine_hours: number | string;
  total_idle_hours: number | string;
  total_fuel_used: number | string;
  utilization_rate: number;
  idle_percentage: number;
  avg_engine_hours_day: number;
  avg_idle_hours_day: number;
  highest_engine_day: DailyUsagePoint | null;
  highest_idle_day: DailyUsagePoint | null;
  business_insight: string;
  active_anomalies: string[];
  daily_trend: DailyUsagePoint[];
  recent_logs: UsageLogResponse[];
}

export interface UtilizationFilterParams {
  equipment_id?: string;
  equipment_type?: string;
  site_id?: string;
  rental_id?: string;
}

export const fetchFleetUtilization = async (params?: UtilizationFilterParams): Promise<EquipmentUtilization[]> => {
  const response = await apiClient.get<EquipmentUtilization[]>('/analytics/utilization', { params });
  return response.data;
};

export const fetchEquipmentUtilization = async (equipmentId: string): Promise<EquipmentUtilization> => {
  const response = await apiClient.get<EquipmentUtilization>(`/analytics/utilization/${encodeURIComponent(equipmentId)}`);
  return response.data;
};

export const fetchSiteAnalytics = async (): Promise<SiteAnalytics[]> => {
  const response = await apiClient.get<SiteAnalytics[]>('/analytics/sites');
  return response.data;
};

export const fetchEquipmentPerformance = async (equipmentId: string): Promise<EquipmentPerformance> => {
  const response = await apiClient.get<EquipmentPerformance>(`/analytics/equipment/${encodeURIComponent(equipmentId)}/performance`);
  return response.data;
};

export const fetchDailyUsageTrend = async (equipmentId: string): Promise<DailyUsagePoint[]> => {
  const response = await apiClient.get<DailyUsagePoint[]>(`/analytics/equipment/${encodeURIComponent(equipmentId)}/daily`);
  return response.data;
};

File: backend/app/schemas/analytics.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import datetime, date
from decimal import Decimal
from app.schemas.usage import UsageLogResponse

class EquipmentUtilization(BaseModel):
    equipment_id: str
    equipment_type: str
    site_id: Optional[str] = None
    site_name: Optional[str] = None
    operator_id: Optional[str] = None
    operator_name: Optional[str] = None
    rental_status: str
    engine_hours: Decimal
    idle_hours: Decimal
    total_usage_hours: Decimal
    fuel_used: Decimal
    utilization_rate: float
    idle_percentage: float
    insight_summary: str

    model_config = ConfigDict(from_attributes=True)

class SiteAnalytics(BaseModel):
    site_id: str
    site_name: str
    location: str
    equipment_count: int
    active_rentals: int
    total_engine_hours: Decimal
    total_idle_hours: Decimal
    total_fuel_used: Decimal
    average_utilization: float

    model_config = ConfigDict(from_attributes=True)

class DailyUsagePoint(BaseModel):
    date: str
    engine_hours: float
    idle_hours: float
    fuel_used: float
    utilization_rate: float

class EquipmentPerformance(BaseModel):
    equipment_id: str
    equipment_type: str
    status: str
    site_id: Optional[str] = None
    site_name: Optional[str] = None
    operator_id: Optional[str] = None
    operator_name: Optional[str] = None
    rental_id: Optional[str] = None
    checkout_date: Optional[datetime] = None
    expected_return: Optional[datetime] = None
    total_engine_hours: Decimal
    total_idle_hours: Decimal
    total_fuel_used: Decimal
    utilization_rate: float
    idle_percentage: float
    avg_engine_hours_day: float
    avg_idle_hours_day: float
    highest_engine_day: Optional[DailyUsagePoint] = None
    highest_idle_day: Optional[DailyUsagePoint] = None
    business_insight: str
    active_anomalies: List[str] = []
    daily_trend: List[DailyUsagePoint] = []
    recent_logs: List[UsageLogResponse] = []
